zapcore/console_encoder.go

```diff
@@ -119,10 +120,23 @@ func (c consoleEncoder) EncodeEntry(ent Entry, fields []Field) (*buffer.Buffer,
 	c.writeContext(line, fields)
 
 	// If there's no stacktrace key, honor that; this allows users to force
-	// single-line output.
+	// single-line output by avoiding printing the stacktrace.
 	if ent.Stack != "" && c.StacktraceKey != "" {
 		line.AppendByte('\n')
-		line.AppendString(ent.Stack)
+
+		arr = getSliceEncoder()
+		stacktraceEncoder := c.EncodeStacktrace
+		if stacktraceEncoder == nil {
+			stacktraceEncoder = FullStacktraceEncoder
+		}
+		stacktraceEncoder(ent.Stack, arr)
+		for i := range arr.elems {
+			if i > 0 {
+				line.AppendString(c.ConsoleSeparator)
+			}
+			fmt.Fprint(line, arr.elems[i])
+		}
```

Comment:
The contract for stacktraceEncoder is that it makes at most one `Append*` call.
So we don't need the full `arr.elems` handling here. It can just be:
```suggestion
		if len(arr.elems) > 0 {
			fmt.Fprint(line, arr.elems[0])
		} else {
			// User-provided encoder was a no-op.
			// Fall back to full encoder.
			line.AppendString(ent.Stack)
		}
```